src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory_QueryLambdaRewriter.vb

```diff
@@ -0,0 +1,93 @@
+                LocalRewriter.PopulateRangeVariableMapForQueryLambdaRewrite(node, _rangeVariableMap, inExpressionLambda:=False)
+                Dim rewrittenBody As BoundStatement = LocalRewriter.RewriteQueryLambdaBody(AddressOf VisitExpressionWithStackGuard, node, _rangeVariableMap)
+                Return LocalRewriter.RewriteQueryLambda(rewrittenBody, node)
+            End Function
+
+            Public Overrides Function VisitRangeVariable(node As BoundRangeVariable) As BoundNode
+                Dim expression As BoundExpression = Nothing
+                If Not _rangeVariableMap.TryGetValue(node.RangeVariable, expression) Then
+                    ' _rangeVariableMap should contain an entry for the range variable, except for error cases. 
+                    Debug.Assert(node.HasErrors OrElse node.RangeVariable.Type.IsErrorType())
+                    Return node
+                End If
+
+#If DEBUG Then
+                ' Range variable reference should be rewritten to a parameter reference, or a call or a property access.
+                ' We clone these bound nodes in QueryLambdaRewriterPass2 to avoid dag in the generated bound tree.
+                ' If the LocalRewriter is changed to generate more kind of bound nodes for range variables, we should handle these in QueryLambdaRewriterPass2.
+                ' Below assert helps us to stay in sync with the LocalRewriter.
+                Select Case expression.Kind
+                    Case BoundKind.Parameter
+                    Case BoundKind.Call
+                    Case BoundKind.PropertyAccess
+                        Exit Select
+                    Case Else
+                        Debug.Fail($"Unexpected bound kind '{expression.Kind}' generated for range variable rewrite by method '{NameOf(LocalRewriter.PopulateRangeVariableMapForQueryLambdaRewrite)}'")
+                End Select
+#End If
+
+                Return expression
+            End Function
+        End Class
+
+        Private NotInheritable Class QueryLambdaRewriterPass2
+            Inherits BoundTreeRewriterWithStackGuard
+            Private ReadOnly _uniqueNodes As HashSet(Of BoundExpression)
+
+            Public Sub New()
+                _uniqueNodes = New HashSet(Of BoundExpression)
+            End Sub
+
+            Private Function HandleNode(Of T As BoundExpression)(node As T, cloneNode As Func(Of T, T)) As T
+                If Not _uniqueNodes.Add(node) Then
+                    node = cloneNode(node)
+                    _uniqueNodes.Add(node)
+                End If
+
+                Return node
+            End Function
+
+            Public Overrides Function VisitParameter(node As BoundParameter) As BoundNode
+                node = DirectCast(MyBase.VisitParameter(node), BoundParameter)
+                Return HandleNode(node, cloneNode:=Function(n) New BoundParameter(node.Syntax, node.ParameterSymbol, node.IsLValue, node.SuppressVirtualCalls, node.Type, node.HasErrors))
```

Comment:
>HandleNode [](start = 23, length = 10)

The code from Handle node can probably be inlined here.

Reply:
Also, we could also check if parameter symbol belongs to a query lambda (ContainingSymbol.IsQueryLambdaMethod) and only check nodes like that here.

---
In reply to: [137415002](https://github.com/dotnet/roslyn/pull/21356#discussion_r137415002) [](ancestors = 137415002)

Reply:
Sounds like a better approach - I will adjust.

---
In reply to: [137446461](https://github.com/dotnet/roslyn/pull/21356#discussion_r137446461) [](ancestors = 137446461,137415002)